### PR TITLE
Expose Started in ConsumerPipeline, nack with requeue true if Consumer cannot handle message

### PR DESCRIPTION
**Pull request by tiggerite**
Also includes some small refactoring of Consumer and use of !.Value to keep the IDE happy.

### CODE REVIEW

**Comment by tristan-hyams** [comment]
What was the issue you were working around with the WaitToRead?

File: src/HouseofCat.RabbitMQ/Consumer/Consumer.cs [hunk 1]
```diff
@@ -84,9 +84,9 @@ await _conLock
                     await SetChannelHostAsync().ConfigureAwait(false);
                     _shutdown = false;
                     _dataBuffer = Channel.CreateBounded<ReceivedData>(
-                        new BoundedChannelOptions(ConsumerOptions.BatchSize.Value)
+                        new BoundedChannelOptions(ConsumerOptions.BatchSize!.Value)
                         {
-                            FullMode = ConsumerOptions.BehaviorWhenFull.Value
+                            FullMode = ConsumerOptions.BehaviorWhenFull!.Value
```
Comment (tristan-hyams) [inline src/HouseofCat.RabbitMQ/Consumer/Consumer.cs:105]: Nice, hadn't come across this mechanism before.

Note to self:
https://docs.microsoft.com/en-us/dotnet/csharp/language-reference/operators/null-forgiving

**Comment by tiggerite** [comment]
Basically, when a microservice talking to rabbit needs to be updated and thus stopped/started, stopping the consumer and associated pipeline isn't enough to prevent dead letters under high load, as there is a period between the last AckMessage call and the ChannelPool being stopped via ShutdownAsync on the RabbitService meaning rabbit still sees the channel and sends the next unacknowledged message to it. However because the consumer is stopped it then can't do anything with it so the message expires and ends up dead lettered.

**Comment by tiggerite** [comment]
It also seems to meet the criteria for

"Sometimes a consumer cannot process a delivery immediately but other instances might be able to. In this case it may be desired to requeue it and let another consumer receive and handle it." as per https://www.rabbitmq.com/confirms.html#consumer-nacks-requeue

**Comment by tiggerite** [comment]
If it feels like too much of a breaking change, I'm happy to take the NackMessage call out and have the event handlers just call HandleMessage without using the return value by default. As long as they're protected virtual I can always create a subclass and override the functionality, now the HandleMessage returns the ReceivedData and a success status.

The only issue is ConsumerDataflow where I can't specify creation of a custom Consumer.. that might need another PR.

**Comment by tiggerite** [comment]
Oh I see you changed it not to return the ReceivedData on failure? But then there is no way to Nack the message from the caller.. the thing is if WaitToWriteAsync returns false or (in a race condition) WriteAsync throws an exception, we still consumed the message due to the BasicConsume and the long lived channel, so we do need a mechanism for nacking it. With this new change that won't be possible though.

**Comment by tristan-hyams** [comment]
In the event a consumer's internal buffer is in an unwriteable state (i.e. exception is thrown), it means we are in a terminal/shutdown state for the consumer.

The Scenario w/ Domain Boundary I think you are stating:

1. User triggers shutdown consumer -> HouseofCat.RabbitMQ.
2. Library shuts consumer down but its not immediate -> Closes internal Channel Write buffer, sets the Consumer also do not reconnect state.
3. Events are still coming from the RabbitMQ library -> messages were still coming in but were ignored (WaitToWrite is false) or a few throw exceptions because WriteAsync will fail. They will eventually be Redelivered by RabbitMQ server side as they will timeout pending an acknowledgement OR the consumer's channel dies the server will release them for consumption.

Consumer shutdown non-immediate
- The "consumer" in my library blocks writing to the HouseofCat.RabbitMQ.Consumer internal memory buffer.
- The "consumption" mechanism from RabbitMQ is still engaged, because closing the channel stops it.
  - We don't want to close the channel as we need it open to ack/nack currently processing messages.
- RabbitMQ.Client is still consuming messages/events but we don't do anything with them (WaitToWrite is false at this point).
  - Assuming one is using durable messages the following is true:
    - There is a delay here before the server attempts redelivery. It's whatever the standard ack/nack timeout is (I think 60 seconds). 
    - It will also only grab the QoS batch count at any one time... so if it's 50, then the server should show 50 are unacked.
    - **Nothing is lost and everything should eventually be processed.**
- The channel is kept open so that all currently processed messages will be acked/nacked based on their usage because they have to be acked/nacked on the same channel as the consumer (a major PITA).
  - Once that channel dies/closes, there is no option but for the server to resend any still pending or unacked messages.
- The Try {} Catch{} only effects a tiny race condition between WaitToWrite being true and then WriteAsync() throws exception, which we were allowing exceptions to be thrown regardless (it doesn't really impact anything and there is nothing that can be done here except maybe debug log/trace them).

There shouldn't be any dead letters or messages lost at all! If there is at this point, there is a major bug somewhere.

Consumer shutdown immedate
- The "consumer" in my library blocks writing to the HouseofCat.RabbitMQ.Consumer internal memory buffer.
- The "consumption" mechanism from RabbitMQ is dead, because closing the channel stops it.
- All messages currently in flight (in internal memory) will be all worked on.
- When it comes time to Ack/Nack, all of them will fail.
   - This is unfortunately unavoidable. All messages will be consumed again Consumer brought online.
   - While I do encourage everything the user does to be "re-runnable" / re-entrant / idempotent etc., this can cause issues of duplicate work.

Regarding (https://www.rabbitmq.com/confirms.html#consumer-nacks-requeue) I think the instruction is misleading, if I have 50 messages currently in process and using QoS batch counts, I will have 50 unacked "pending processing internally". Depending on what you are doing with the messages, this could take a mere few milliseconds or maybe 50 seconds in total CPU time to process all 50 (obviously in parallel). The concept of nacking a message to send to another consumer while simultaneously using batch counts are mutually exclusive. You can't buffer AND nack! They made a logical mistake.

Lastly, the purpose of the Consumer is deliver data to another area (asynchronously) to perform work. That area will also eventually Ack/Nack. If it wasn't for the stupid Channel reference needed, there would be no linkage whatsoever bringing it back to the Consumer. The life-cycle of the message should be controlled by user, more so than the library (i.e. you know when you are done with it) barring technical issues -> which is really all my library is for convenience and handling the technical gotchas. That bit of code you have added for Nacking while definitely possible to occur for an exception (it's a slim race condition) it really isn't preventing anything nor stopping a dead letter scenario on the server.

A lot of this though, is on the assumption of using Durable Messages + No Server outages occurred.

**Comment by tristan-hyams** [comment]
> Oh I see you changed it not to return the ReceivedData on failure? But then there is no way to Nack the message from the caller.. the thing is if WaitToWriteAsync returns false or (in a race condition) WriteAsync throws an exception, we still consumed the message due to the BasicConsume and the long lived channel, so we do need a mechanism for nacking it. With this new change that won't be possible though.

That's correct, I am still writing the next part though :), I will edit my comments above for the full homogenous blob of text.

**Comment by tiggerite** [comment]
Thanks for the long explanation, your RabbitMQ knowledge is certainly better than mine :) fair enough on the not needing to nack, then. We might as well remove that code altogether because if handled is false then rabbitMessage will be null and it'll blow up now?

**Comment by tristan-hyams** [comment]
Already took care of it, its good to know this stuff. Learn from my failures ^.^ It would have helped if I had written comments.... (crickets).

Everything else looks good as of now, I will cut a release with these other changes and my QoL changes.

**Comment by tiggerite** [comment]
I'll see if I can track down the cause of the dead letters in the meantime. The microservice in question can take a long time to process a message (up to 30s) and it's on a kubernetes pod, so in theory gets a grace period where I should be able to ack the messages in flight as long as I can stop any more coming in - the replacement service starts consuming straightaway when it is considered "live" so it should take over. It also publishes messages just to add to the fun!

**Comment by tristan-hyams** [comment]
I think you wanted to create ConsumerDataflows but with Custom Consumers. While that was and still is the intention, I think the mechanism of having to go and build a custom RabbitService in order to use those Custom Consumers was well meaning but clearly a lot of extra work... so I have added a constructor to bypass that for the time being by passing in a list of IConsumers and some additional properties I think it will cover your use case, but let me know.

> I'll see if I can track down the cause of the dead letters in the meantime. The microservice in question can take a long time to process a message (up to 30s) and it's on a kubernetes pod, so in theory gets a grace period where I should be able to ack the messages in flight as long as I can stop any more coming in - the replacement service starts consuming straightaway when it is considered "live" so it should take over. It also publishes messages just to add to the fun!

Try setting Debug/Trace visibility level on for logger. There maybe enough additional detail to see what's happening.

In addition to that, you can (server side) configure a Dead Letter Exchange/Queue. You can maybe find the messages/types that are going bye bye and maybe reverse engineer the problem.

**Comment by tiggerite** [comment]
> I think you wanted to create ConsumerDataflows but with Custom Consumers. While that was and still is the intention, I think the mechanism of having to go and build a custom RabbitService in order to use those Custom Consumers was well meaning but clearly a lot of extra work... so I have added a constructor to bypass that for the time being by passing in a list of IConsumers and some additional properties I think it will cover your use case, but let me know.

This sounds perfect. Thanks so much!
 
> Try setting Debug/Trace visibility level on for logger. There maybe enough additional detail to see what's happening.

I actually set Verbose and yes it's very chatty. The only problem is getting the logs off the terminating pod. You have to be quick as once it's killed off they're gone for good.. I should really set up a local version but the load test is properly beefy. It throws around 6 messages per second at the service I think (which has three pods and each pod two connections, as the rabbit is clusteted).

> In addition to that, you can (server side) configure a Dead Letter Exchange/Queue. You can maybe find the messages/types that are going bye bye and maybe reverse engineer the problem.

Yep, this we have already, but it only ever tells me "expired". Not too helpful really..

**Comment by tristan-hyams** [comment]
I will await the results! Keep me posted.

**Comment by tiggerite** [comment]
By the way everything you wrote about the non immediate shutdown is exactly our scenario. But unfortunately having a message timeout isn't helpful for the use case (because it has to be an immediate response) and this is why it becomes a dead letter, I believe. 

I have another idea about nacking for requeue outside of your (amazing btw!) library now, which as you rightly say is the place for it, hopefully it will mitigate the situation. I prefer ConsumerPipeline in general by the way (it's quicker, less memory exhaustive, and easier to follow the code) but for a few of our services Dataflow is required simply because of the consumer count - it overwhelms Kestrel otherwise as each pipeline requires a thread pool.

**Comment by tiggerite** [comment]
I _think_ I have solved the dead letter problem by keeping track of the delivered counter and waiting up to 60s on shutdown for that to reduce to 0, via Interlocked trickery to not need actual locks/semaphores which would slow down normal message processing. It also stops individual consumers with immediate set as soon as that consumer counter reaches 0. Finally a bit of nacking with requeue for catching those newly received during the small window of counter being >0 but shutdown being triggered.

**Pull request merged**